doc/compiler/optimizer/ValuePropagation.md

```diff
@@ -0,0 +1,195 @@
+
+In summary, merging constraints only makes the information known at the time worse, or at the same level as before.
+
+## Examples of Constraints
+
+### Range Constraints
+
+These constraints represent ranges of values a particular value number can take at a given point in the program. As we saw previously these ranges can be discrete values, or true continuous (potentially discrete) ranges. Range constraints can be used to eliminate control flow or propagate particular values to different points in the program. There are many range constraint classes for tracking different types, such as integers and longs.
+
+### Type Constraints
+
+For typed languages it is often beneficial to flow type constraints. Knowing type information about a particular value number at a given point in the program can allow us to devirtualize calls or eliminate type check and exception edges.
+
+There are three different constraint classes which flow this information in VP, they are the [fixed class](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L655-L674) constraint, [resolved class](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L628-L653) constraint, and [unresolved class](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L706-L736) constraints.
+
+Fixed class constraints simply express that a value number has a specific, fixed, type `T`. We can know this information perhaps because we have seen the allocation of the value number expression or perhaps we have seen an explicit cast into type `T`. Knowing the exact type of a value number expression is the most useful.
+
+Resolved class constraints express that a value number is of type `T` or any subclass of `T`. These constraints are useful for eliminating additional type checks on the value number at some later point in the program.
+
+Unresolved class constraints are least interesting and express that the class type of the value number has not been loaded by the runtime and is unknown. They simply carry around the string representing the class of the object.
+
+### Null and NonNull Constraints
+
+There are two notions flowed through VP for reference types in regards to nullness. The [null](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L746-L760) constraint, and the [non-null](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L762-L775) constraint. These constraints are useful for tracking value numbers which are certainly known to be either `null` or certainly not `null`. Knowing such information can eliminate implicit and explicit null checks or open doors for further simplifications.
+
```

Comment:
These also help fold away comparisons (ifacmpeq/ne) that test against null.

Reply:
Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).